File: hash/leetcode_17/solution.js
/**
 * @param {string} digits
 * @return {string[]}
 */

const letterCombinations = (digits) => {
  if (digits == null || digits.length === 0) return [];

  const map = ['','','abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz' ]
  const result = [];

  const backtrack = (index, value) => {
    if (index === digits.length) {
      result.push(value);
      return;
    }

    for (const char of map[digits[index]]) {
      backtrack(index + 1, value + char);
    }
  };

  backtrack(0, '');

  return result;
};

File: hash/leetcode_17/solution.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
            if not digits:
                return []

            map =  ['','','abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz' ]
            result = []

            def backtrack(index, value):
                if index == len(digits):
                    result.append(value)
                    return

                letters = map[int(digits[index])]
                
                for char in letters:
                    backtrack(index + 1, value + char)

            backtrack(0, "")

            return result